samples/aspnetcore/blazor/FlightFinder/FlightFinder.Client/Services/AppState.cs

```diff
@@ -27,25 +29,57 @@ public AppState(HttpClient httpInstance)
             http = httpInstance;
         }
 
-        public async Task Search(SearchCriteria criteria)
+        public async Task SearchFlights(FlightSearchCriteria criteria)
         {
             SearchInProgress = true;
             NotifyStateChanged();
 
-            SearchResults = await http.PostJsonAsync<Itinerary[]>("/api/flightsearch", criteria);
+            FlightSearchResults = await http.PostJsonAsync<FlightItinerary[]>("/api/flightsearch", criteria);
             SearchInProgress = false;
             NotifyStateChanged();
         }
 
-        public void AddToShortlist(Itinerary itinerary)
+        public async Task SearchRentalCars(RentalCarSearchCriteria criteria)
+        {
+            SearchInProgress = true;
+            NotifyStateChanged();
+
+            RentalCarSearchResults = await http.PostJsonAsync<RentalCarItinerary[]>("/api/rentalcarsearch", criteria);
+            SearchInProgress = false;
+            NotifyStateChanged();
+        }
+
+        public void AddToShortlist(FlightItinerary itinerary)
+        {
+            var item = new Components.Shortlist.Item();
+            item.Content = (builder) =>
+            {
+                builder.OpenComponent<Components.ShortlistFlightItem>(0);
+                builder.AddAttribute(1, "Itinerary", itinerary);
+                builder.AddAttribute(2, "OnRemoveItinerary", () => RemoveFromShortlist(item));
+                builder.CloseComponent();
+            };
+            shortlist.Add(item);
+            NotifyStateChanged();
```

Comment:
I don't love this.... I want something like WPF's data templates to solve this - I know we've talked about this in the past. Any ideas of a better short term solution?

Reply:
Although it adds a further `if`/`else` case to a component, how about:

Instead of putting component instances (or `RenderFragment` instances) within the `AppState` structure, rather keep that purely as state information. For example, make both `FlightItinerary` and `RentalCarItinerary` inherit from some abstract `Itinerary`, and have the type of `ShortList` still be `IReadOnlyList<Itinerary>`.

Then in `Shortlist.cshtml`:

```html
<h2>Shortlist (@Items.Count)</h2>

@foreach (var item in Items)
{
    <ul class="list-group mb-4">
        @if (item is FlightItinerary flightItinerary)
        {
            <ShortlistFlightItem
                Itinerary=flightItinerary
                OnRemoveItinerary=@(() => OnRemoveItinerary(flightItinerary)) />
        }
        else if (item is RentalCarItinerary rentalCarItinerary)
        {
            <ShortlistCarRentalItem
                Itinerary=rentalCarItinerary
                OnRemoveItinerary=@(() => OnRemoveItinerary(rentalCarItinerary)) />
        }
    </ul>
}

@functions
{
    [Parameter] IReadOnlyList<Itinerary> Items { get; set; }
    [Parameter] Action<Itinerary> OnRemoveItinerary { get; set; }
}
```